lang_GENERIC/analyze/unit_typing_java.ml

```diff
@@ -0,0 +1,70 @@
+open Common
+open OUnit
+module V = Visitor_AST
+module A = AST_generic
+
+(*****************************************************************************)
+(* Unit tests *)
+(*****************************************************************************)
+
+let unittest =
+  "typing_java" >::: [
+
+    "test basic variable definitions" >:: (fun () ->
+      let file = Filename.concat Config_pfff.path "/tests/GENERIC/typing/VarDef.java" in
+        try
+          let ast = Parse_generic.parse_program file in
+            let lang = List.hd (Lang.langs_of_filename file) in 
+             Naming_AST.resolve lang ast;
+
+            let v = V.mk_visitor { V.default_visitor with
+                V.kexpr = (fun (_k, _) exp ->
+                    match exp with
+                      | A.Id(_, {A.id_type=id_type; _}) -> (
+                            match !id_type with 
+                              | Some(A.TyName((("String", _)), _)) -> ()
+                              | _ -> assert_failure("Variable referenced did not have expected type String"))
+                      | _ -> ()
+                );
+            } in
+            v (A.Pr ast)
+        with Parse_info.Parsing_error _ ->
+          assert_failure (spf "it should correctly parse %s" file)
+    );
+
+    "test multiple variable definitions" >:: (fun () ->
+      let file = Filename.concat Config_pfff.path "/tests/GENERIC/typing/EqVarCmp.java" in
+        try
+          let ast = Parse_generic.parse_program file in
+            let lang = List.hd (Lang.langs_of_filename file) in 
+             Naming_AST.resolve lang ast;
+
+            let v = V.mk_visitor { V.default_visitor with
+                V.kexpr = (fun (_k, _) exp ->
+                  match exp with
+                    | A.Call(_, x::y::[]) ->
+                      ((match x with
+                        | A.Arg(A.Id(_, {A.id_type=id_type; _})) -> (
```

Comment:
You can also just match A.Id (("var1", _, {a.id_type = ...})) to know that you're on "var1".
No need to match the A.Call and A.Arg context I think.